mvrx/src/main/kotlin/com/airbnb/mvrx/MavericksViewModelConfigFactory.kt

```diff
@@ -0,0 +1,89 @@
+package com.airbnb.mvrx
+
+import android.content.Context
+import android.content.pm.ApplicationInfo
+import androidx.annotation.RestrictTo
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Factory for providing the [MavericksViewModelConfig] for each new ViewModel that is created.
+ *
+ * An instance of this must be set on [MvRx.viewModelConfigFactory].
+ *
+ * A custom subclass of this may be used to allow you to override [buildConfig], but this should
+ * generally not be necessary.
+ */
+open class MavericksViewModelConfigFactory(
+    /**
+     * True if debug checks should be run. Should be false for production builds.
+     * When true, certain validations are applied to the ViewModel. These can be slow and should
+     * not be used in production! However, they do help to catch common issues so it is highly
+     * recommended that you enable debug when applicable.
+     */
+    val debugMode: Boolean,
+    /**
+     * Provide a default context for viewModelScope. It will be added after [SupervisorJob]
+     * and [Dispatchers.Main.immediate].
+     */
+    val contextOverride: CoroutineContext? = null
+) {
+
+    /**
+     * Sets [debugMode] depending on whether the app was built with the Debuggable flag enabled.
+     */
+    constructor(context: Context) : this(context.isDebuggable())
+
+    private val onConfigProvidedListener =
+        mutableListOf<(BaseMavericksViewModel<*>, MavericksViewModelConfig<*>) -> Unit>()
+
+    open fun coroutineScope(): CoroutineScope {
+        return CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate + contextOverride)
+    }
+
+    internal fun <S : MvRxState> provideConfig(
+        viewModel: BaseMavericksViewModel<S>,
+        initialState: S
+    ): MavericksViewModelConfig<S> {
+        return buildConfig(viewModel, initialState).also { config ->
+            onConfigProvidedListener.forEach { callback -> callback(viewModel, config) }
+        }
+    }
+
+    /**
+     * Create a new [MavericksViewModelConfig] for the given viewmodel.
+     * This can be overridden to customize the config.
+     */
+    open fun <S : MvRxState> buildConfig(
+        viewModel: BaseMavericksViewModel<S>,
+        initialState: S
+    ): MavericksViewModelConfig<S> {
+        return object : MavericksViewModelConfig<S>(debugMode, CoroutinesStateStore(initialState, coroutineScope()), coroutineScope()) {
```

Comment:
Is it intentional that the store and view model have different scopes?

Reply:
nope, good catch